src/libraries/System.Linq/src/System/Linq/Shuffle.SpeedOpt.cs

```diff
@@ -0,0 +1,272 @@
+                        // reservoir with a decreasing probability.
+                        long totalElementsSeen = reservoir.Count;
+                        while (e.MoveNext())
+                        {
+                            long r = Random.Shared.NextInt64(totalElementsSeen);
+                            if (r < reservoir.Count)
+                            {
+                                reservoir[(int)r] = e.Current;
+                            }
+                        }
+                    }
+                }
+
+            ReturnReservoir:
+                if (reservoir is not null)
+                {
+                    // Ensure that elements in the reservoir are in random order. The sampling helped
+                    // to ensure we got a uniform distribution from the source into the reservoir, but
+                    // it didn't randomize the order of the reservoir itself; this is especially relevant
+                    // to the elements initially added into the reservoir.
+                    Random.Shared.Shuffle(CollectionsMarshal.AsSpan(reservoir));
+                }
+
+                return reservoir;
+            }
+
+            public override TSource[] ToArray() => SampleToList(_source, _takeCount)?.ToArray() ?? [];
+
+            public override List<TSource> ToList() => SampleToList(_source, _takeCount) ?? [];
+
+            public override int GetCount(bool onlyIfCheap) =>
+                !onlyIfCheap ? Math.Min(_takeCount, _source.Count()) :
+                TryGetNonEnumeratedCount(_source, out int count) ? Math.Min(_takeCount, count) :
+                -1;
+
+            public override TSource? TryGetFirst(out bool found) =>
+                TryGetElementAt(0, out found);
+
+            public override TSource? TryGetLast(out bool found) =>
+                TryGetElementAt(0, out found);
+
+            public override TSource? TryGetElementAt(int index, out bool found)
+            {
+                if (_source is Iterator<TSource> iterator &&
+                    iterator.GetCount(onlyIfCheap: true) is int iteratorCount &&
+                    iteratorCount >= 0)
+                {
+                    if ((uint)index < (uint)Math.Min(_takeCount, iteratorCount))
+                    {
+                        return iterator.TryGetElementAt(Random.Shared.Next(0, iteratorCount), out found);
+                    }
+                }
+                else if (_source is IList<TSource> list)
+                {
+                    int count = list.Count;
+                    if ((uint)index < (uint)Math.Min(_takeCount, count))
+                    {
+                        found = true;
+                        return list[Random.Shared.Next(0, count)];
+                    }
+                }
+                else if (index >= 0)
+                {
+                    TSource[] array = _source.ToArray();
```

Comment:
How often is TryGetElementAt expected to be called? ToArray() on every index operation seems quite expensive to me

Reply:
It's no more expensive han if this wasn't here, in which case ElementAt which still do this. It's similar to Reverse().ElementAt on an enumerable today, or OrderBy(). ElementAt. Hopefully it's not frequently used. That said, now that I have the sample method, I can use that here to take away a bit of the sting, though it'll still be O(n).

Reply:
I think my implied question was "should this do `_source = _source.ToArray()` if this ever gets hit?", but perhaps not.

Reply:
> I think my implied question was "should this do `_source = _source.ToArray()` if this ever gets hit?", but perhaps not.

That would break the typical LINQ assumption that every iteration is fresh.

I've changed the implementation to not create the full array each time, though it does still need to fully enumerate.